Further reduction of allocs in mutating methods for mixtures

PerezHz opened the pull request:
During the development of #347, @lbenet suggested a way to further reduce allocations in mutating division methods for mixtures (see [this comment](https://github.com/JuliaDiff/TaylorSeries.jl/pull/347#discussion_r1501682516)). After running some local tests, it indeed looks like an option worthwhile pursuing.

On current master:
```julia
julia> using BenchmarkTools

julia> using TaylorSeries
Precompiling TaylorSeries
  1 dependency successfully precompiled in 3 seconds. 3 already precompiled.

julia> dq = set_variables("δq", order=1, numvars=2)
2-element Vector{TaylorN{Float64}}:
  1.0 δq₁ + 𝒪(‖x‖²)
  1.0 δq₂ + 𝒪(‖x‖²)

julia> x = Taylor1([2+2dq[1],-1.1dq[1], 0.7dq[2], 0.5dq[1]-0.45dq[2],0.9dq[1]])
  2.0 + 2.0 δq₁ + 𝒪(‖x‖²) + ( - 1.1 δq₁ + 𝒪(‖x‖²)) t + ( 0.7 δq₂ + 𝒪(‖x‖²)) t² + ( 0.5 δq₁ - 0.45 δq₂ + 𝒪(‖x‖²)) t³ + ( 0.9 δq₁ + 𝒪(‖x‖²)) t⁴ + 𝒪(t⁵)

julia> z = deepcopy(x)
  2.0 + 2.0 δq₁ + 𝒪(‖x‖²) + ( - 1.1 δq₁ + 𝒪(‖x‖²)) t + ( 0.7 δq₂ + 𝒪(‖x‖²)) t² + ( 0.5 δq₁ - 0.45 δq₂ + 𝒪(‖x‖²)) t³ + ( 0.9 δq₁ + 𝒪(‖x‖²)) t⁴ + 𝒪(t⁵)

julia> TaylorSeries.zero!(z)

julia> iszero(z)
true

julia> z[0] = 1/x[0]
 0.5 - 0.5 δq₁ + 𝒪(‖x‖²)

julia> @btime TaylorSeries.div!($z, 1, $x, 1)
  735.992 ns (18 allocations: 1.27 KiB)

julia> @btime TaylorSeries.div!($z, 1, $x, 2)
  1.342 μs (36 allocations: 2.53 KiB)

julia> @btime TaylorSeries.div!($z, 1, $x, 3)
  2.000 μs (54 allocations: 3.80 KiB)

julia> @btime TaylorSeries.div!($z, 1, $x, 4)
  2.657 μs (72 allocations: 5.06 KiB)
```
This PR:
```julia
julia> @btime TaylorSeries.div!($z, 1, $x, 1)
  685.610 ns (18 allocations: 1.27 KiB)

julia> @btime TaylorSeries.div!($z, 1, $x, 2)
  708.029 ns (18 allocations: 1.27 KiB)

julia> @btime TaylorSeries.div!($z, 1, $x, 3)
  724.354 ns (18 allocations: 1.27 KiB)

julia> @btime TaylorSeries.div!($z, 1, $x, 4)
  740.625 ns (18 allocations: 1.27 KiB)
```
~~The remaining 18 allocations are coming from this line:
https://github.com/JuliaDiff/TaylorSeries.jl/blob/bc1735c32546415c1c254a342537c28c361f988b/src/arithmetic.jl#L815~~
EDIT: the afore-mentioned 18 allocations are not there anymore.

UPDATE: after successfully reducing allocation in mutating division and product methods for mixtures, these changes enabled further allocation reduction in sqrt, pow, and other functions. These improvements have also been added to this PR.

lbenet commented on src/arithmetic.jl:
```diff
@@ -530,14 +592,14 @@ c_k = \sum_{j=0}^k a_j b_{k-j}.
 """
 @inline function mul!(c::HomogeneousPolynomial, a::HomogeneousPolynomial,
-        b::HomogeneousPolynomial)
+        b::HomogeneousPolynomial, d::T=1) where {T<:NumberNotSeries}
```
I do see the interest to have this method. Yet, I wonder if it does induce some sort of type unstability... My point is, since the default value is `1::Int`, if we set it to something else (say 1.0), do we use the same method or a different one? Perhaps it is better to have two methods, `mul!(c, a, b)` and `mul!(c, a, b, d::T)`, where the latter includes multiplying by the "scalar" `d::T<:NumberNotSeries`.

PerezHz replied:
I agree there's some potential for type-instability here; this way it just seemed really convenient to avoid duplicating code. I'll see if I can find another solution.

PerezHz replied:
Regarding this point, the current approach is to add `scalar::NumberNotSeries` as a keyword argument, which avoids dispatching on this argument. Adding a `mul!(a, b, c, d::T)` can call the wrong method of `mul!` when `d` is an integer. Another approach can also be to simply define `mul_scalar!(a, b, c, d::NumberNotSeries)` and call this method where appropriate. What are your thoughts on this?

lbenet replied:
I favor your the proposal with `mul_scalar!`, simply because it will have less methods than `mul!`...

lbenet commented on src/arithmetic.jl:
```diff
@@ -823,6 +890,8 @@ end
         return nothing
     end
 
+    zero!(c, k)
+
```
I think that `zero(c, k)` makes that `div!` does not accumulate anymore the result. So I suggest to correct the note. 

Another suggestion would be to have two methods `div_accumul!` and `div` (and `mul_accumul!` and `mul!`), and use them properly. I think this would simplify `div!` below a lot...

PerezHz replied:
> I think that zero(c, k) makes that div! does not accumulate anymore the result. So I suggest to correct the note.

Good catch! Okay, I'll fix it 👍

lbenet commented on src/arithmetic.jl:
```diff
@@ -832,6 +901,63 @@ end
+#
+# Recursion algorithm:
+#
+# k = 0: c[0] <-  c[0]/a[0]
+# k = 1: c[1] <- -c[1]
+#        c[1] <-  c[1] + c[0]*a[1]
+#        c[1] <- -c[1]/a[0]
+# k = 2: c[2] <- -c[2]
+#        c[2] <-  c[2] + c[0]*a[2] + c[1]*a[1]
+#        c[2] <- -c[2]/a[0]
```
This resembles `fma`... maybe we should (in another PR) implement it

PerezHz replied:
I agree it's worth implementing!

lbenet commented:
@PerezHz I have left three comments on what you have so far in this PR; not sure if they are good ideas or not. In any case, this looks very promising!

PerezHz commented:
> @PerezHz I have left three comments on what you have so far in this PR; not sure if they are good ideas or not. In any case, this looks very promising!

Many thanks @lbenet! This is still WIP but even so, many thanks for the comments you left; I'll try to address them in the next few days! I very much agree the changes here can be improved 😄

PerezHz commented:
@lbenet I managed to get rid of almost all allocations in `pow!`, except for this line:
https://github.com/PerezHz/TaylorSeries.jl/blob/f95b797a00766396f46f378416129875dbe7528a/src/power.jl#L299
I'm thinking a possible solution could be to simply define a mutating method `power_by_squaring!` for `TaylorN`, waht do you think? Or maybe there's a better way?

lbenet commented:
Thanks again, truly, this is looking awesome!!

> @lbenet I managed to get rid of almost all allocations in `pow!`, except for this line: https://github.com/PerezHz/TaylorSeries.jl/blob/f95b797a00766396f46f378416129875dbe7528a/src/power.jl#L299 I'm thinking a possible solution could be to simply define a mutating method `power_by_squaring!` for `TaylorN`, what do you think? Or maybe there's a better way?

While I think it is possible to reduce the number of allocations with respect to the current implementation, the implementation (seems to me) not trivial, because of the recursive nature of the function and the fact that products (and squares) are computed order-by-order. I thus suggest you give it a try, with the possibility to do it in another PR.

PerezHz commented:
> While I think it is possible to reduce the number of allocations with respect to the current implementation, the implementation (seems to me) not trivial, because of the recursive nature of the function and the fact that products (and squares) are computed order-by-order. I thus suggest you give it a try, with the possibility to do it in another PR.

This is something I'm still looking at, but perhaps we can leave it as a to-do for the time being. In the end, one allocation per a full expansion is still way better than it was before. I'm still tackling a few remaining cases, but this should be ready for review in the next couple of days...

lbenet commented:
Fully agree...

PerezHz commented:
Actually, I think the changes here so far are in a sense self-contained and by themselves achieve good performance improvements. If you agree I'll leave further changes for another PR, so this is now ready for review 😄.

lbenet commented on src/functions.jl:
```diff
@@ -303,10 +303,35 @@ end
-        @inline function identity!(c::$T{T}, a::$T{T}, k::Int) where {T<:Number}
-            @inbounds c[k] = identity(a[k])
+        @inline function identity!(c::$T{T}, a::$T{T}, k::Int) where {T<:NumberNotSeries}
+            if $T == Taylor1
+                @inbounds c[k] = identity(a[k])
+            else
+                @inbounds for l in eachindex(c[k])
+                    identity!(c[k], a[k], l)
+                end
+            end
```
This `if` is within a `for` loop, which I guess is bad for performance. I suggest to have it the other way around, though I admit that it implies duplicating the `for` loops in both `if` branches...

PerezHz replied:
Thanks for spotting this, I'm now changing the occurrences of this pattern, although this case in particular has the `for` within the `else` so this one is actually correct, right?

lbenet replied:
Ups... sorry

lbenet commented on src/functions.jl:
```diff
@@ -345,19 +370,18 @@ for T in (:Taylor1, :TaylorN)
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = k * a[k] * c[0]
-            else
-                @inbounds mul!(c[k], k * a[k], c[0])
-            end
-            @inbounds for i = 1:k-1
+            @inbounds for i = 0:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * a[k-i] * c[i]
                 else
-                    mul!(c[k], (k-i) * a[k-i], c[i])
+                    mul!(c[k], a[k-i], c[i], scalar=k-i)
                 end
             end
-            @inbounds c[k] = c[k] / k
+            if $T == Taylor1
+                @inbounds div!(c, c, k, k)
+            else
+                @inbounds div!(c[k], c[k], k)
+            end
```
Same comment as above; in this case, I think the second `if`s can be included in the first.

PerezHz replied:
Done.

lbenet commented on src/functions.jl:
```diff
@@ -371,16 +395,20 @@ for T in (:Taylor1, :TaylorN)
             end
             @inbounds for i = 1:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * a[k-i] * c[i]
                 else
-                    mul!(c[k], (k-i) * a[k-i], c[i])
+                    mul!(c[k], a[k-i], c[i], scalar=k-i)
                 end
             end
-            @inbounds c[k] = c[k] / k
+            if $T == Taylor1
+                @inbounds div!(c, c, k, k)
+            else
+                @inbounds div!(c[k], c[k], k)
+            end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done.

lbenet commented on src/functions.jl:
```diff
@@ -393,16 +421,11 @@ for T in (:Taylor1, :TaylorN)
-            end
-            @inbounds for i = 2:k-1
+            @inbounds for i = 1:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * a[i] * c[k-i]
                 else
-                    mul!(c[k], (k-i)*a[i], c[k-i])
+                    mul!(c[k], a[i], c[k-i], scalar=k-i)
                 end
             end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done.

lbenet commented on src/functions.jl:
```diff
@@ -423,16 +446,11 @@ for T in (:Taylor1, :TaylorN)
-            end
-            @inbounds for i = 2:k-1
+            @inbounds for i = 1:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * a[i] * c[k-i]
                 else
-                    mul!(c[k], (k-i)*a[i], c[k-i])
+                    mul!(c[k], a[i], c[k-i], scalar=k-i)
                 end
             end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done.

lbenet commented on src/functions.jl:
```diff
@@ -442,70 +460,79 @@ for T in (:Taylor1, :TaylorN)
+                    @inbounds s[0][1], c[0][1] = sincos( a0 )
+                end
                 return nothing
             end
-            x = a[1]
             zero!(s, k)
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds s[k] = x * c[k-1]
-                @inbounds c[k] = -x * s[k-1]
-            else
-                mul!(s[k], x, c[k-1])
-                mul!(c[k], -x, s[k-1])
-            end
-            @inbounds for i = 2:k
-                x = i * a[i]
+            @inbounds for i = 1:k
                 if $T == Taylor1
+                    x = i * a[i]
                     s[k] += x * c[k-i]
                     c[k] -= x * s[k-i]
                 else
-                    mul!(s[k], x, c[k-i])
-                    mul!(c[k], -x, s[k-i])
+                    mul!(s[k], a[i], c[k-i], scalar=i)
+                    mul!(c[k], a[i], s[k-i], scalar=-i)
                 end
             end
 
-            @inbounds s[k] = s[k] / k
-            @inbounds c[k] = c[k] / k
+            if $T == Taylor1
+                s[k] = s[k] / k
+                c[k] = c[k] / k
+            else
+                @inbounds div!(s[k], s[k], k)
+                @inbounds div!(c[k], c[k], k)
+            end
             return nothing
         end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -442,70 +460,79 @@ for T in (:Taylor1, :TaylorN)
+            sincos!(s, c, a, k)
             return nothing
         end
 
         @inline function tan!(c::$T{T}, a::$T{T}, c2::$T{T}, k::Int) where {T<:Number}
             if k == 0
                 @inbounds aux = tan( constant_term(a) )
-                @inbounds c[0] = aux
-                @inbounds c2[0] = aux^2
+                if $T == Taylor1
+                    @inbounds c[0] = aux
+                    @inbounds c2[0] = aux^2
+                else
+                    @inbounds c[0][1] = aux
+                    @inbounds c2[0][1] = aux^2
+                end
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = k * a[k] * c2[0]
-            else
-                @inbounds mul!(c[k], k * a[k], c2[0])
-            end
-            @inbounds for i = 1:k-1
+            @inbounds for i = 0:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * a[k-i] * c2[i]
                 else
-                    mul!(c[k], (k-i) * a[k-i], c2[i])
+                    mul!(c[k], a[k-i], c2[i], scalar=k-i)
                 end
             end
-            @inbounds c[k] = a[k] + c[k]/k
+            # c[k] <- c[k]/k
+            if $T == Taylor1
+                div!(c, c, k, k) # Taylor1
+            else
+                div!(c[k], c[k], k) # TaylorN
+            end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -514,25 +541,48 @@ for T in (:Taylor1, :TaylorN)
         @inline function asin!(c::$T{T}, a::$T{T}, r::$T{T}, k::Int) where {T<:Number}
             if k == 0
                 a0 = constant_term(a)
-                @inbounds c[0] = asin( a0 )
-                @inbounds r[0] = sqrt( 1 - a0^2 )
+                if $T == Taylor1
+                    @inbounds c[0] = asin( a0 )
+                    @inbounds r[0] = sqrt( 1 - a0^2 )
+                else
+                    @inbounds c[0][1] = asin( a0 )
+                    @inbounds r[0][1] = sqrt( 1 - a0^2 )
+                end
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = (k-1) * r[1] * c[k-1]
-            else
-                @inbounds mul!(c[k], (k-1) * r[1], c[k-1])
-            end
-            @inbounds for i in 2:k-1
+            @inbounds for i in 1:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * r[i] * c[k-i]
                 else
-                    mul!(c[k], (k-i) * r[i], c[k-i])
+                    mul!(c[k], r[i], c[k-i], scalar=k-i)
+                end
+            end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -514,25 +541,48 @@ for T in (:Taylor1, :TaylorN)
+            # Compute auxiliary term s=1-a^2
+            s = ($T)(zero(a[0]), a.order)
+            one_s = one(s)
+            for i = 0:k
+                sqr!(s, a, i)
+                subst!(s, s, i)
+                if i == 0
+                    if $T == Taylor1
+                        s[0] = one(s[0]) - s[0]
+                    else
+                        s[0][1] = one(s[0][1]) - s[0][1]
+                    end
+                    add!(s, one_s, s, 0)
+                end
+            end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

PerezHz replied:
(actually the code changed here a bit so this `for` is not there anymore)

lbenet commented on src/functions.jl:
```diff
@@ -569,16 +614,11 @@ for T in (:Taylor1, :TaylorN)
                 return nothing
             end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -569,16 +614,11 @@ for T in (:Taylor1, :TaylorN)
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = (k-1) * r[1] * c[k-1]
-            else
-                @inbounds mul!(c[k], (k-1) * r[1], c[k-1])
-            end
-            @inbounds for i in 2:k-1
+            @inbounds for i in 1:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * r[i] * c[k-i]
                 else
-                    mul!(c[k], (k-i) * r[i], c[k-i])
+                    mul!(c[k], r[i], c[k-i], scalar=k-i)
                 end
             end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -595,14 +635,7 @@ for T in (:Taylor1, :TaylorN)
-                @inbounds mul!(c[k], x, s[k-1])
-            end
-            @inbounds for i = 2:k
+            @inbounds for i = 1:k
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -625,16 +658,11 @@ for T in (:Taylor1, :TaylorN)
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = k * a[k] * c2[0]
-            else
-                @inbounds mul!(c[k], k * a[k], c2[0])
-            end
-            @inbounds for i = 1:k-1
+            @inbounds for i = 0:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * a[k-i] * c2[i]
                 else
-                    mul!(c[k], (k-i) * a[k-i], c2[i])
+                    mul!(c[k], a[k-i], c2[i], scalar=k-i)
                 end
             end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -651,16 +679,11 @@ for T in (:Taylor1, :TaylorN)
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = (k-1) * r[1] * c[k-1]
-            else
-                @inbounds mul!(c[k], (k-1) * r[1], c[k-1])
-            end
-            @inbounds for i in 2:k-1
+            @inbounds for i in 1:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * r[i] * c[k-i]
                 else
-                    mul!(c[k], (k-i) * r[i], c[k-i])
+                    mul!(c[k], r[i], c[k-i], scalar=k-i)
                 end
             end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -676,16 +699,11 @@ for T in (:Taylor1, :TaylorN)
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = (k-1) * r[1] * c[k-1]
-            else
-                @inbounds mul!(c[k], (k-1) * r[1], c[k-1])
-            end
-            @inbounds for i in 2:k-1
+            @inbounds for i in 1:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * r[i] * c[k-i]
                 else
-                    mul!(c[k], (k-i) * r[i], c[k-i])
+                    mul!(c[k], r[i], c[k-i], scalar=k-i)
                 end
             end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/functions.jl:
```diff
@@ -701,16 +719,11 @@ for T in (:Taylor1, :TaylorN)
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = (k-1) * r[1] * c[k-1]
-            else
-                @inbounds mul!(c[k], (k-1) * r[1], c[k-1])
-            end
-            @inbounds for i in 2:k-1
+            @inbounds for i in 1:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * r[i] * c[k-i]
                 else
-                    mul!(c[k], (k-i) * r[i], c[k-i])
+                    mul!(c[k], r[i], c[k-i], scalar=k-i)
                 end
             end
```
Same comment about order of `if` and `for`

PerezHz replied:
Done 👍

lbenet commented on src/arithmetic.jl:
```diff
@@ -453,28 +507,53 @@ end
 for T in (:Taylor1, :TaylorN)
     # NOTE: For $T = TaylorN, `mul!` *accumulates* the result of a * b in c[k]
-    @eval @inline function mul!(c::$T{T}, a::$T{T}, b::$T{T}, k::Int) where {T<:Number}
+    @eval @inline function mul!(c::$T{T}, a::$T{T}, b::$T{T}, k::Int; scalar::NumberNotSeries=1) where {T<:Number}
```
Eventhough I think `mul_scalar!` is better in this context (too many `mul!` methods, I find it ok to have the keyword argument.

PerezHz replied:
Thanks for this comment, I agree that it's better not to "over-overload" `mul!` so actually in the end I did add `mul_scalar!` and also the equivalent `div_scalar!`.

lbenet reviewed (commented):
I've left a bunch of comments, mainly in src/functions.jl; they are about inverting (outer) `for` loops, and `if` blocks. Other than that, LGTM.

PerezHz commented:
> I've left a bunch of comments, mainly in src/functions.jl; they are about inverting (outer) for loops, and if blocks. Other than that, LGTM.

Thank you for spotting this! I've pushed some changes addressing your comments, I'm guessing this time we can bump a patch version, since only the internal API is changing? Tests on TaylorIntegration and NEOs pass with this branch without any changes 👍

lbenet commented on src/arithmetic.jl:
```diff
@@ -469,12 +523,53 @@ for T in (:Taylor1, :TaylorN)
+        else
+            @inbounds mul_scalar!(c[k], scalar, a[0], b[k])
+        end
+        @inbounds for i = 1:k
+            if $T == Taylor1
+                c[k] += a[i] * b[k-i]
+            else
+                mul_scalar!(c[k], scalar, a[i], b[k-i])
+            end
+        end
```
Same comment wrt order of `for` and `if`

lbenet replied:
This can be rewritten as
```julia
if $T == Taylor1
    @inbounds c[k] = a[0] * b[k]
    @inbounds for i = 1:k
        c[k] += a[i] * b[k-i]
     end
else
    @inbounds mul_scalar!(c[k], scalar, a[0], b[k])
    @inbounds for i = 1:k
        mul_scalar!(c[k], scalar, a[i], b[k-i])
    end
end
```

lbenet commented on src/functions.jl:
```diff
@@ -442,70 +459,83 @@ for T in (:Taylor1, :TaylorN)
+                end
+            end
+            # c[k] <- c[k]/k
+            if $T == Taylor1
+                div!(c, c, k, k) # Taylor1
+            else
+                div!(c[k], c[k], k) # TaylorN
             end
```
This `if` can be merged to the previous one

lbenet commented on src/functions.jl:
```diff
@@ -596,24 +653,24 @@ for T in (:Taylor1, :TaylorN)
+                    mul_scalar!(c[k], i, a[i], s[k-i])
+                end
+            end
+            if $T == Taylor1
+                @inbounds div!(s, s, k, k)
+                @inbounds div!(c, c, k, k)
+            else
+                @inbounds div!(s[k], s[k], k)
+                @inbounds div!(c[k], c[k], k)
```
This if block can also be merged with the previous one

lbenet commented on src/functions.jl:
```diff
@@ -626,18 +683,21 @@ for T in (:Taylor1, :TaylorN)
-                    c[k] += (k-i) * a[k-i] * c2[i]
-                else
-                    mul!(c[k], (k-i) * a[k-i], c2[i])
+            if $T == Taylor1
+                @inbounds c[k] = a[k] - c[k]/k
+            else
+                @inbounds for l in eachindex(c[k])
+                    c[k][l] = a[k][l] - c[k][l]/k
```
Same comment: `if`s can be merged

lbenet commented:
> Thank you for spotting this! I've pushed some changes addressing your comments, I'm guessing this time we can bump a patch version, since only the internal API is changing? Tests on TaylorIntegration and NEOs pass with this branch without any changes 👍

I've left few more comments along the same lines as before. In any case, I do agree with you that we should bump a patch version before merging this PR. It's also very nice that tests pass smoothly in TaylorIntegration.jl and NEOs.jl; have you checked the performance?

PerezHz commented:
> have you checked the performance?

So far I have found ~40% improvements in memory usage and performance, sometimes even better!

lbenet commented:
> So far I have found ~40% improvements in memory usage and performance, sometimes even better!

Very impressive! Please, let me know when you think this is ready, and I'll go ahead and merge it, and bump the patch version.

PerezHz commented:
> Very impressive! Please, let me know when you think this is ready, and I'll go ahead and merge it, and bump the patch version.

Just rebased to current master, and unless CI complains, this is ready to be merged!

lbenet commented:
Thanks a lot again @PerezHz. I'll wait for the tests to complete, and then merge it!

lbenet commented:
Merging!
